doc/man/polybar-msg.1.rst
polybar-msg(1)
==============

SYNOPSIS
--------
| **polybar-msg** [*OPTIONS*] **action** *action-string*
| **polybar-msg** [*OPTIONS*] **action** *module* *action* [*data*]
| **polybar-msg** [*OPTIONS*] **cmd** *command*

DESCRIPTION
-----------
Polybar allows external control through *actions* and *commands*.
Actions control individual modules and commands control the bar itself.

The full IPC documentation is linked at the end of this document.

The available actions depend on the target module.
For actions, the payload is either a single action string or the module name,
the action name, and the optional data string specified separately.

In order for **polybar-msg** being able to send a message to a running
**polybar** process, the bar must have IPC enabled and both **polybar-msg** and
**polybar** must run under the same user.

OPTIONS
-------

.. program:: polybar-msg

.. option:: -h, --help

   Display help text and exit

.. option:: -p PID

   Send message only to **polybar** process running under the given process ID.
   If not specified, the message is sent to all running **polybar** processes.

EXAMPLES
--------

**polybar-msg** **cmd** *quit*
  Terminate all running **polybar** instances.

**polybar-msg** **action** *mymodule* *module_hide*

**polybar-msg** **action** "*#mymodule.module_hide*"
  Hide the module named *mymodule*.
  The first variant specifies the module and action names separately, the second uses an action string.

AUTHORS
-------
| Polybar was created by Michael Carlberg and is currently maintained by Patrick Ziegler.
| Contributors can be listed on GitHub.

REPORTING BUGS
--------------
Report issues on GitHub <https://github.com/polybar/polybar>

SEE ALSO
--------
.. only:: man

  :manpage:`polybar`\(1),
  :manpage:`polybar`\(5)

  | IPC documentation: <https://polybar.rtfd.org/en/stable/user/ipc.html>


.. only:: not man

  :doc:`polybar.1`,
  :doc:`polybar.5`

  :doc:`/user/ipc`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
Each release should have the following subsections, if entries exist, in the
given order: `Breaking`, `Build`, `Deprecated`, `Removed`, `Added`, `Changed`,
`Fixed`, `Security`.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Breaking
- We added the backslash escape character (\\) for configuration values. This
  means that the literal backslash character now has special meaning in
  configuration files, therefore if you want to use it in a value as a literal
  backslash, you need to escape it with the backslash escape character. The
  parser logs an error if any unescaped backslashes are found in a value. This
  affects you only if you are using two consecutive backslashes in a value,
  which will now be interpreted as a single literal backslash.
  [`#2354`](https://github.com/polybar/polybar/issues/2354)
- We rewrote our tag parser. This shouldn't break anything, if you experience
  any problems, please let us know.
  The new parser now gives errors for certain invalid tags where the old parser
  would just silently ignore them. Adding extra text to the end of a valid tag
  now produces an error. For example, tags like `%{T-a}`, `%{T2abc}`, `%{rfoo}`,
  and others will now start producing errors.
  This does not affect you unless you are producing your own formatting tags
  (for example in a script) and you are using one of these invalid tags.
- For security reasons, the named pipe at `/tmp/polybar_mqueue.<PID>` had its
  permission bits changed from `666` to `600` to prevent sending ipc messages
  to polybar processes running under a different user.

### Build
- New dependency: [libuv](https://github.com/libuv/libuv). At least version 1.3
  is required.
- Bump the minimum cmake version to 3.5
- The `BUILD_IPC_MSG` option has been renamed to `BUILD_POLYBAR_MSG`
- Building the documentation is now enabled by default and not just when
  `sphinx-build` is found.
- Users can control exactly which targets should be available with the following
  cmake options (together with their default value):
  - `BUILD_POLYBAR=ON` - Builds the `polybar` executable
  - `BUILD_POLYBAR_MSG=ON` - Builds the `polybar-msg` executable
  - `BUILD_TESTS=OFF` - Builds the test suite
  - `BUILD_DOC=ON` - Builds the documentation
  - `BUILD_DOC_HTML=BUILD_DOC` - Builds the html documentation (depends on `BUILD_DOC`)
  - `BUILD_DOC_MAN=BUILD_DOC` - Builds the manpages (depends on `BUILD_DOC`)
  - `BUILD_CONFIG=ON` - Generates the default config
  - `BUILD_SHELL=ON` - Generates shell completion files
  - `DISABLE_ALL=OFF` - Disables all above targets by default. Individual
    targets can still be enabled explicitly.
- The documentation can no longer be built by directly configuring the `doc`
  directory.
- The sample config file is now placed in the `generated-sources` folder inside
  whatever folder you invoked `cmake` from instead of in the root folder of the
  repository.
- The `POLYBAR_FLAGS` cmake variable can be used to pass extra C++ compiler flags.
- The sample config file has been removed.
- Polybar now ships a default config that is installed to
  `/etc/polybar/config.ini`, it lives in `doc/config.ini`.
  It will also be placed in the `examples` directory in the documentation folder.
  [`#2405`](https://github.com/polybar/polybar/issues/2405)
- The `userconfig` target has been removed, you can no longer use `make
  userconfig`. As an alternative, you can copy the default config from
  `/etc/polybar/config.ini`.

### Deprecated
- `[settings]`: `throttle-output` and `throttle-output-for` have been removed.
  The new event loop already does a similar thing where it coalesces update
  triggers if they happen directly after one another, leading to only a single
  bar update.
- When not specifying the config file with `--config`, naming your config file
  `config` is deprecated. Rename your config file to `config.ini`.
- Directly writing ipc messages to `/tmp/polybar_mqueue.<PID>` is deprecated,
  users should always use `polybar-msg`. As a consequence the message format
  used for IPC is deprecated as well.
- `polybar-msg hook` is deprecated in favor of using the hook action.
  `polybar-msg` will tell you the correct command to use.

### Removed
- `DEBUG_SHADED` cmake variable and its associated functionality.

### Added
- Right and middle click events for alsa module.
  ([`#2566`](https://github.com/polybar/polybar/issues/2566))
- `internal/network`: New token `%mac%` shows MAC address of selected interface
   ([`#2568`](https://github.com/polybar/polybar/issues/2568))
- Polybar can now read config files from stdin: `polybar -c /dev/stdin`.
- `custom/script`: Implement `env-*` config option.
   ([`#2090`](https://github.com/polybar/polybar/issues/2090))
- `drawtypes/ramp`: Add support for ramp weights.
   ([`#1750`](https://github.com/polybar/polybar/issues/1750))
- `internal/memory`: New tokens `%used%`, `%free%`, `%total%`, `%swap_total%`,
  `%swap_free%`, and `%swap_used%` that automatically switch between MiB and GiB
  when below or above 1GiB.
  ([`#2472`](https://github.com/polybar/polybar/issues/2472))
- Option to always show urgent windows in i3 module when `pin-workspace` is active
  ([`#2374`](https://github.com/polybar/polybar/issues/2374))
- `internal/xworkspaces`: `reverse-scroll` can be used to reverse the scroll
  direction when cycling through desktops.
- The backslash escape character (\\).
  [`#2354`](https://github.com/polybar/polybar/issues/2354)
- Warn states for the cpu, memory, fs, and battery modules.
  ([`#570`](https://github.com/polybar/polybar/issues/570),
  [`#956`](https://github.com/polybar/polybar/issues/956),
  [`#1871`](https://github.com/polybar/polybar/issues/1871),
  [`#2141`](https://github.com/polybar/polybar/issues/2141))

src/polybar-msg.cpp
#include <fcntl.h>
#include <unistd.h>

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <cstring>
#include <deque>

#include "common.hpp"
#include "components/eventloop.hpp"
#include "ipc/decoder.hpp"
#include "ipc/encoder.hpp"
#include "ipc/msg.hpp"
#include "ipc/util.hpp"
#include "modules/ipc.hpp"
#include "utils/actions.hpp"
#include "utils/file.hpp"

using namespace std;
using namespace polybar;
using namespace eventloop;

static const char* exec = nullptr;
static constexpr auto USAGE = "<command=(action|cmd)> <payload> [...]";
static constexpr auto USAGE_HOOK = "hook <module-name> <hook-index>";

void display(const string& msg) {
  fprintf(stdout, "%s\n", msg.c_str());
}

void error(const string& msg) {
  throw std::runtime_error(msg);
}

void uv_error(int status, const string& msg) {
  throw std::runtime_error(msg + " (" + uv_strerror(status) + ")");
}

void usage(FILE* f, const string& parameters) {
  fprintf(f, "Usage: %s [-p pid] %s\n", exec, parameters.c_str());
}

void remove_socket(const string& handle) {
  if (unlink(handle.c_str()) == -1) {
    error("Could not remove stale ipc channel: "s + strerror(errno));
  } else {
    display("Removed stale ipc channel: " + handle);
  }
}

bool validate_type(const string& type) {
  return (type == "action" || type == "cmd" || type == "hook");
}

static vector<string> get_sockets() {
  auto sockets = file_util::glob(ipc::get_glob_socket_path());

  auto new_end = std::remove_if(sockets.begin(), sockets.end(), [](const auto& path) {
    int pid = ipc::get_pid_from_socket(path);

    if (pid <= 0) {
      return true;
    }

    if (!file_util::exists("/proc/" + to_string(pid))) {
      remove_socket(path);
      return true;
    }

    return false;
  });

  sockets.erase(new_end, sockets.end());

  return sockets;
}

static void on_write(PipeHandle& conn, ipc::decoder& dec) {
  conn.read_start(
      [&](const auto& e) {
        try {
          if (!dec.closed()) {
            dec.on_read(reinterpret_cast<const uint8_t*>(e.data), e.len);
          }
        } catch (const ipc::decoder::error& e) {
          conn.close();
        }
      },
      [&]() { conn.close(); },
      [&](const auto& e) {
        conn.close();
        uv_error(e.status, "There was an error while reading polybar's response");
      });
}

static void on_connection(PipeHandle& conn, ipc::decoder& dec, const ipc::type_t type, const string& payload) {
  const auto data = ipc::encode(type, payload);
  conn.write(
      data, [&]() { on_write(conn, dec); },
      [&](const auto& e) {
        conn.close();
        uv_error(e.status, "There was an error while sending the IPC message.");
      });
}

static std::pair<ipc::type_t, string> parse_message(deque<string> args) {
  // Validate args
  const string ipc_type{args.front()};
  args.pop_front();
  string ipc_payload{args.front()};
  args.pop_front();

  if (!validate_type(ipc_type)) {
    error("\"" + ipc_type + "\" is not a valid message type.");
  }

  ipc::type_t type = ipc::TYPE_ERR;

  /*
   * Check hook specific args
   *
   * The hook type is deprecated. Its contents are translated into a hook action.
   */
  if (ipc_type == "hook") {
    if (args.size() != 1) {
      usage(stderr, USAGE_HOOK);
      throw std::runtime_error("Mismatched number of arguments for hook, expected 1, got "s + to_string(args.size()));
    } else {
      if (ipc_payload.find("module/") == 0) {
        ipc_payload.erase(0, strlen("module/"));
      }

      // Hook commands use 1-indexed hooks but actions use 0-indexed ones
      int hook_index = std::stoi(args.front()) - 1;
      args.pop_front();

      type = to_integral(ipc::v0::ipc_type::ACTION);
      ipc_payload =
          actions_util::get_action_string(ipc_payload, polybar::modules::ipc_module::EVENT_HOOK, to_string(hook_index));

      fprintf(stderr,
          "Warning: Using IPC hook commands is deprecated, use the hook action on the ipc module: %s %s \"%s\"\n", exec,
          ipc_type.c_str(), ipc_payload.c_str());
    }
  }

  if (ipc_type == "action") {
    type = to_integral(ipc::v0::ipc_type::ACTION);
    /**
     * Alternatively polybar-msg action <module name> <action> <data>
     * is also accepted
     */
    if (!args.empty()) {
      string name = ipc_payload;
      string action = args.front();
      args.pop_front();
      string data{};
      if (!args.empty()) {
        data = args.front();
        args.pop_front();
      }

      ipc_payload = actions_util::get_action_string(name, action, data);
    }
  }

  if (ipc_type == "cmd") {
    type = to_integral(ipc::v0::ipc_type::CMD);
  }

  if (!args.empty()) {
    error("Too many arguments");
  }

  assert(type != ipc::TYPE_ERR);
  return {type, ipc_payload};
}

int run(int argc, char** argv) {
  deque<string> args{argv + 1, argv + argc};

  string socket_path;

  auto help_pos = find_if(args.begin(), args.end(), [](const string& a) { return a == "-h" || a == "--help"; });
  if (help_pos != args.end()) {
    usage(stdout, USAGE);
    return EXIT_SUCCESS;
  }

  /* If -p <pid> is passed, check if the process is running and that
   * a valid channel socket is available
   */
  if (args.size() >= 2 && args[0].compare(0, 2, "-p") == 0) {
    auto& pid_string = args[1];
    socket_path = ipc::get_socket_path(pid_string);
    if (!file_util::exists("/proc/" + pid_string)) {
      error("No process with pid " + pid_string);
    } else if (!file_util::exists(socket_path)) {
      error("No channel available for pid " + pid_string);
    }

    args.pop_front();
    args.pop_front();
  }

  // If no pid was given, search for all open sockets.
  auto sockets = socket_path.empty() ? get_sockets() : vector<string>{socket_path};

  // Get availble channel sockets
  if (sockets.empty()) {
    error("No active ipc channels");
  }

  if (args.size() < 2) {
    usage(stderr, USAGE);
    return EXIT_FAILURE;
  }

  string payload;
  ipc::type_t type;
  std::tie(type, payload) = parse_message(args);
  string type_str = type == to_integral(ipc::v0::ipc_type::ACTION) ? "action" : "command";

  bool success = true;

  loop loop;

  logger null_logger{loglevel::NONE};

  /*
   * Store all decoreds in vector so that they're alive for the whole eventloop.
   */
  vector<ipc::decoder> decoders;

  for (auto&& channel : sockets) {
    int pid = ipc::get_pid_from_socket(channel);
    assert(pid > 0);

    decoders.emplace_back(
        null_logger, [pid, channel, &payload, &type_str, &success](uint8_t, ipc::type_t type, const auto& response) {
          switch (type) {
            case ipc::TYPE_OK:
              printf("Successfully wrote %s '%s' to PID %d\n", type_str.c_str(), payload.c_str(), pid);
              break;
            case ipc::TYPE_ERR: {
              string err_str{response.begin(), response.end()};
              fprintf(stderr, "%s: Failed to write %s '%s' to PID %d (reason: %s)\n", exec, type_str.c_str(),
                  payload.c_str(), pid, err_str.c_str());
              success = false;
              break;
            }
            default:
              fprintf(stderr, "%s: Got back unrecognized message type %d from PID %d\n", exec, type, pid);
              success = false;
              break;
          }
        });

    /*
     * Index to decoder is captured because reference can be invalidated due to the vector being modified.
     */
    int idx = decoders.size() - 1;

    auto& conn = loop.handle<PipeHandle>();
    conn.connect(
        channel,
        [&conn, &decoders, type, payload, channel, idx]() { on_connection(conn, decoders[idx], type, payload); },
        [&](const auto& e) {
          fprintf(stderr, "%s: Failed to connect to '%s' (err: '%s')\n", exec, channel.c_str(), uv_strerror(e.status));
          success = false;
        });
  }

  try {
    loop.run();
  } catch (const exception& e) {
    throw std::runtime_error("Uncaught exception in eventloop: "s + e.what());
  }

  return success ? EXIT_SUCCESS : EXIT_FAILURE;
}

int main(int argc, char** argv) {
  exec = argv[0];
  try {
    return run(argc, argv);
  } catch (const std::exception& e) {
    fprintf(stderr, "%s: %s\n", exec, e.what());
    return EXIT_FAILURE;
  }
}
